Implement Orchard Bundle Validation for Zcash Withdrawals

r-near opened the pull request:
## What This Does

Adds security validation for Orchard (shielded) withdrawals in the Zcash bridge. Users can now withdraw to Unified Addresses with proper enforcement that bundles go to the right recipient with the right amount.

## The Problem

The `orchard_tx` branch added basic Orchard transaction support, but had TODOs around validation:
- No verification that Orchard outputs go to the intended recipient
- No verification that amounts match what the user is withdrawing
- No protection against bundle tampering after transparent signatures are collected

This meant a malicious user could potentially redirect shielded funds or manipulate amounts.

## The Solution

### 1. OVK-Based Output Recovery

Implemented `orchard_policy.rs` module that uses a hardcoded Bridge OVK to recover and validate Orchard outputs:

```rust
pub const BRIDGE_OVK: [u8; 32] = [0u8; 32];

pub fn validate_orchard_bundle(
    bundle: &Bundle<Authorized, ZatBalance>,
    expected_recipient: &str,
    expected_amount: u64,
    chain: &Chain,
) {
    // Recover output using Bridge OVK
    let (recovered_amount, recovered_addr) = recover_orchard_output(bundle);

    // Validate matches expectations
    require!(recovered_amount == expected_amount, "Orchard amount mismatch");
    require!(recovered_addr == expected_addr_bytes, "Orchard recipient mismatch");
}
```

**Trade-off:** The OVK is public (contract code is public), so anyone can recover these outputs. This is acceptable for the bridge use case - we're enforcing policy, not providing privacy.

### 2. ZIP-244 Sighash Binding

The Orchard bundle is now included in the `TransactionData` used for transparent signature generation. This means:
- Any modification to the Orchard bytes invalidates all transparent signatures
- Prevents post-signing bundle swap attacks


### 3. Orchard-Only Withdrawals

Added support for withdrawals with zero transparent outputs (all funds → shielded pool):

```rust
if psbt.get_output_num() == 0 {
    // Orchard-only case: skip transparent validation
    let gas_fee = max_gas_fee.unwrap_or(10000);
    let orchard_amount = amount - withdraw_fee - gas_fee;
    (orchard_amount, gas_fee)
} else {
    // Normal transparent output validation
    self.check_withdraw_psbt_valid(...)
}
```

### 4. Unified Address Validation

Enforces that Unified Addresses must come with Orchard bundles:

```rust
if target_btc_address.starts_with("u1") || target_btc_address.starts_with("u") {
    require!(orchard_bundle.is_some(),
        "Unified Address provided without Orchard bundle. \
         Either provide both or use a transparent Zcash address (starts with t1/t3)");
}
```

## Test Coverage

Added 5 comprehensive tests covering:
- Happy path e2e withdrawal with OVK validation
- Amount mismatch rejection
- Recipient mismatch rejection (different spending keys → different recipients)
- Unified Address without bundle rejection
- Bundle serialization in final transaction

**Note:** Tests cache generated Orchard bundles because proof generation takes ~80-90s. First run is slow, subsequent runs are fast (~16s total).

```bash
# Build first
cargo near build non-reproducible-wasm --features zcash --out-dir res --no-abi

# Run tests
cargo test -p satoshi-bridge --features zcash \
  --test test_orchard_validation \
  --test test_orchard_withdrawal \
  -- --test-threads=1 --nocapture
```

## Breaking Changes

### Event Encoding
Changed `SignedBtcTransaction` event from hex to base64 to stay under NEAR's 16KB log limit:

```diff
- tx_bytes: &'a Vec<u8>
+ tx_bytes_base64: String
```

Overhead reduced from 2x (hex) to 1.33x (base64). Relayer will need to update the parser.

r-near committed:
feat: zcash validation

olga24912 commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -84,22 +84,34 @@ impl Contract {
+        // For Orchard-only withdrawals (no transparent output), skip transparent validation
+        let (actual_received_amount, gas_fee) = if psbt.get_output_num() == 0 {
+            // Orchard-only case: all funds go to shielded pool
+            let gas_fee = max_gas_fee.map(|g| g.0).unwrap_or(10000);
```
The formula is more complex than that. 10,000 is just the minimum gas amount — it applies only to the case where we have at most two inputs and one Orchard output. If, for example, you have 10 inputs, the fee will be different.



https://github.com/zcash/zips/blob/main/zips/zip-0317.rst

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/pull/20/commits/148dfaa4e7c9f8493e85217328bdc62e28daadf4

olga24912 commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -84,22 +84,34 @@ impl Contract {
+        // For Orchard-only withdrawals (no transparent output), skip transparent validation
+        let (actual_received_amount, gas_fee) = if psbt.get_output_num() == 0 {
+            // Orchard-only case: all funds go to shielded pool
+            let gas_fee = max_gas_fee.map(|g| g.0).unwrap_or(10000);
```
Also, I’m not sure whether max_gas_fee will make sense in this case.

olga24912 commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -84,22 +84,34 @@ impl Contract {
-        );
+
+        // For Orchard-only withdrawals (no transparent output), skip transparent validation
+        let (actual_received_amount, gas_fee) = if psbt.get_output_num() == 0 {
```
Perhaps we should isolate this part of the code so that it applies only to the ZCash case.

r-near replied:
All Orchard features are now conditional on `orchard_bundle` presence or ZCash chain checks

olga24912 commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -84,22 +84,34 @@ impl Contract {
+            (orchard_amount, gas_fee)
+        } else {
+            // Transparent output case: validate transparent outputs
+            let target_address_script_pubkey = self
```
The most classic case is when the target address is a unified address that contains an Orchard address but not a transparent address, yet the change is sent to a transparent output.

In this situation, the program will crash with an error.

r-near replied:
Yes, you're right. I do think we should explicitly reject mixed scenarios for simplicity. Like require the change to also be Orchard?

r-near replied:
Added validation here to reject: https://github.com/Near-One/btc-bridge/pull/20/commits/9ce848d3001496bbae293fb2ad54c03848f9d1b7

olga24912 replied:
But we can’t reject mixed scenarios… Our change address is transparent. We have to handle such cases.

olga24912 replied:
Our baseline scenario is one Orchard output and one transparent change.

r-near replied:
Ok I changed the validation here: https://github.com/Near-One/btc-bridge/pull/20/commits/c8995a78e0de11069a06ed3368626122d66eda98

olga24912 commented on contracts/satoshi-bridge/src/chain_signature.rs:
```diff
@@ -164,12 +164,19 @@ impl Contract {
                     account_id: &account_id,
                     tx_id: btc_pending_sign_id.clone(),
-                    tx_bytes: &tx_bytes_with_sign,
+                    tx_bytes_base64,
```
Can we do this only for ZCash but not for Bitcoin, so that we don’t break backward compatibility for Bitcoin?

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/pull/20/commits/91fd74a8352213c4eb96dba39d51792ae973a438

olga24912 replied:
Add a feature flag

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/d068f3f13e6ca1dd2637d7d67948cf28a80836e1

kiseln replied:
What are the consequences if we do break it? Would be nice to have the same behaviour

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -97,12 +99,30 @@ impl Contract {
+            let withdraw_fee = self.internal_config().withdraw_bridge_fee.get_fee(amount.0);
+            // The orchard amount is the withdrawal amount minus the withdraw fee and gas fee
+            // max_gas_fee defaults to 10000 satoshis if not provided
+            let gas_fee = max_gas_fee.map(|g| g.0).unwrap_or(10000);
```
10,000 is the wrong value, see the comment above. And we should create a separate function for this.

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/pull/20/commits/0933d9d982ea2d762a947f3f55128445208e30b0

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -167,6 +193,15 @@ impl Contract {
         orchard_bundle: Option<Vec<u8>>,
     ) -> PromiseOrValue<U128> {
+        // Validate: Unified Address requires Orchard bundle
+        if target_btc_address.starts_with("u1") || target_btc_address.starts_with("u") {
```
This is not a correct check. We may have Unified Addresses from which transparent addresses are extracted. This is a valid scenario.

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/pull/20/commits/b92d1ba64abb364040418bf81890ecc3c845fe84

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -64,17 +71,31 @@ impl PsbtWrapper {
+            } else {
+                // If no expected values provided, at minimum enforce single action
+                require!(
+                    bundle.actions().len() == 1,
```
And in what case could this be valid? If we have a bundle, then they should provide us with the expected addr and amt, right?

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,91 @@
+) {
+    // Enforce single action
+    require!(
+        bundle.actions().len() == 1,
```
Looks like the MIN amount of Actions is 2: https://github.com/zcash/orchard/blob/main/src/builder.rs#L36

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/pull/20/commits/fa3490e48b38baefc9fbcc632e3418ca509f6d0f

r-near committed:
fix: use ZIP-317 formula with orchard action count for fee calculation

Previously, get_min_fee() hardcoded 0 for orchard_action_count when
calling the ZIP-317 fee_required function. This resulted in incorrect
fee calculations for transactions with Orchard bundles.

Now properly counts actions from the orchard bundle (if present) and
passes this to the ZIP-317 FeeRule. The formula accounts for:
- Transparent inputs (P2PKH standard size)
- Transparent outputs (serialized size)
- Orchard actions (actual count from bundle)

The fee calculation varies with transaction structure per ZIP-317
specification - 10,000 satoshis is just the minimum for simple
transactions with minimal inputs/outputs.

r-near committed:
fix: replace hardcoded gas fee with ZIP-317 calculation

Previously used hardcoded 10000 satoshis as default gas fee. Now:

1. Calculate actual fee using ZIP-317 formula via psbt.get_min_fee()
2. Use max_gas_fee as an upper bound when provided
3. Fall back to computed fee when max_gas_fee not provided

This properly accounts for transaction structure (inputs, outputs,
orchard actions) per ZIP-317 spec. The fee varies with tx complexity
rather than using a fixed minimum value.

Changes in contract_methods.rs:
- Create preliminary PSBT to compute fee before validation
- Recreate PSBT with correct expected amounts for validation

Changes in token_receiver.rs:
- Compute ZIP-317 fee and use max_gas_fee as cap

r-near committed:
fix: update minimum orchard actions check from 1 to 2

Per Orchard protocol specification, bundles must contain at least 2
actions for privacy (MIN_ACTIONS = 2). The Orchard builder
automatically pads bundles with dummy actions to meet this minimum.

Changes:
- Added MIN_ACTIONS constant (2) in orchard_policy.rs
- Updated validation to require >= MIN_ACTIONS instead of == 1
- Applied to both strict validation (with expected values) and
  looser validation (internal operations)
- Improved error messages to show expected vs actual action count

Reference: https://github.com/zcash/orchard/blob/main/src/builder.rs#L36

r-near committed:
fix: use base64 encoding only for ZCash to preserve Bitcoin compatibility

Previously, all chains used base64 encoding for tx_bytes in the
SignedBtcTransaction event, which breaks backward compatibility
for Bitcoin indexers expecting hex encoding.

Now:
- ZCash chains (mainnet/testnet): Use base64 (1.33x overhead)
  - Beneficial for larger Orchard transactions
- Bitcoin/other chains: Use hex encoding (backward compatible)

The event field name remains `tx_bytes_base64` but contains hex
for non-ZCash chains to avoid breaking the event structure.

r-near committed:
fix: validate Unified Address for actual Orchard receiver presence

Previously, the code checked if an address starts with "u1" or "u"
to determine if an Orchard bundle is required. This is incorrect
because Unified Addresses can contain only transparent receivers
(P2PKH/P2SH), which is a valid scenario without an Orchard bundle.

Now:
- Added has_orchard_receiver() helper to check if a Unified Address
  actually contains an Orchard receiver
- Only require Orchard bundle if the UA has an Orchard receiver
- Allow Unified Addresses with only transparent receivers to work
  without an Orchard bundle

Transparent addresses can be extracted from Unified Addresses,
which is a valid use case.

r-near committed:
fix: reject mixed Orchard and transparent output scenarios

Previously, the code would crash when a Unified Address with only
an Orchard receiver was provided along with transparent change outputs.
The transparent validation path would fail trying to extract a
transparent script_pubkey from an Orchard-only UA.

Now:
- Added has_orchard_bundle() helper to PsbtWrapper
- Explicitly reject transactions with both Orchard bundle and
  transparent outputs
- Require either pure Orchard (no transparent outputs) or pure
  transparent (no Orchard bundle)

This prevents the classic case where target has Orchard receiver
but change goes to transparent output, which would cause the
program to crash.

The restriction simplifies validation logic and avoids edge cases.
Users must choose one path or the other for clarity.

olga24912 commented on contracts/satoshi-bridge/src/psbt.rs:
```diff
@@ -2,6 +2,7 @@ use crate::network::Address;
 use crate::psbt_wrapper::PsbtWrapper;
 use crate::*;
 impl Contract {
+    #[allow(clippy::too_many_arguments)]
```
We can’t perform check_withdraw_psbt because we don’t know the Orchard output.
Maybe we should deserialize the amount from the Orchard data and use it for validation?

r-near replied:
Good point, added it here: https://github.com/Near-One/btc-bridge/pull/20/commits/5d4c90d4c640ec54a1d9dc117d58372be5499425

r-near committed:
fix: validate actual Orchard output amount from bundle

Previously, for Orchard-only withdrawals, the code calculated what the output amount *should* be but didn't validate the actual amount in the Orchard bundle. This was a critical security issue - users could provide any amount in their Orchard bundle.

Now:
- Added get_orchard_output_amount() to PsbtWrapper that recovers the actual output amount using the bridge OVK
- Validate that the recovered Orchard amount is within the expected range (accounting for fees and potential dust adjustment)
- Use the same validation logic as transparent outputs: allow a range between (expected - min_change_amount) and expected

The solution deserializes and validates the Orchard output amount, ensuring users receive the correct amount after fees.

r-near committed:
fix: support mixed Orchard output with transparent change (baseline case)

Previously, commit 9ce848d3001496bbae293fb2ad54c03848f9d1b7 rejected mixed scenarios (Orchard + transparent),
but this was incorrect. The baseline scenario for Orchard withdrawals is:
- 1 Orchard output to the user's Unified Address
- 1+ transparent change outputs back to the bridge

Now properly supports mixed scenarios:
- Validate the Orchard output amount (to user)
- Validate all transparent outputs are change to bridge address
- Ensure change amounts are within min/max limits
- Verify change + gas_fee doesn't exceed input

The new validation logic:
- Orchard bundle present: Validate Orchard output + transparent change
- No Orchard bundle: Use existing transparent-only validation

r-near committed:
fix: comprehensive Orchard bundle validation improvements

Fixed 4 critical validation gaps identified during code review:

1. **Value Balance Validation** (NEW)
   - Now validates bundle.value_balance() equals -(output_amount)
   - Ensures value flows correctly from transparent to Orchard pool
   - Prevents malicious bundles with incorrect value balance

2. **All Outputs Validated** (NEW)
   - Iterates through ALL actions/outputs (not just first)
   - Ensures exactly 1 non-zero output exists
   - Prevents hiding extra outputs in unused actions

3. **Range Validation for Dust Adjustment** (FIXED)
   - Changed from exact match to range validation
   - Allows user to adjust amount for dust (like transparent does)
   - Range: [expected - min_change_amount, expected]
   - Provides same flexibility as transparent withdrawals

4. **Exact Transparent Accounting** (FIXED)
   - Changed `<=` to `==` in transparent change validation
   - Validates: input == orchard_amount + change + gas_fee
   - Prevents missing value in accounting
   - Ensures value balance integrity

These fixes provide defense-in-depth. While the ZCash network would
catch most issues, we now validate comprehensively at contract level
to reject bad requests early and ensure robust accounting.

olga24912 committed:
tmp contract fix

olga24912 commented on contracts/satoshi-bridge/src/kdf.rs:
```diff
@@ -3,6 +3,29 @@ use crate::*;
 
+#[near]
+impl Contract {
+    pub fn generate_public_key_pub(&self, path: String) -> String {
```
A function useful for Bridge-SDK-RS. Overall, you can do without it, but having it makes things more convenient.

We need to get rid of the duplicated code in the function below and possibly replace path with tx_id or something like that.

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/11cd30aea49664d1385121cf9877697a8f1ef5b9

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -34,7 +35,7 @@ macro_rules! define_rbf_callback {
                 #[callback_unwrap] last_block_height: u32,
             ) {
-                let expiry_height = last_block_height + self.get_config().expiry_height_gap;
+                let expiry_height = 0;//last_block_height + self.get_config().expiry_height_gap;
```
Temporarily replaced expire_height with zero, because we need to know expire_height in advance. Probably, as a solution, we should pass expiry_height and then check that it falls within a certain window.

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/21cf32667d6a60f4ec6dfce301f9aab3716e8b9f

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -322,23 +378,28 @@ impl PsbtWrapper {
 fn get_branch_id(expiry_height: u32, config: &Config) -> BranchId {
-    let current_height = expiry_height - config.expiry_height_gap;
-    config.chain.get_branch_id(current_height)
+    BranchId::Nu6_1
```
Locally the logic is correct, but in the future we’ll need to be able to update to new versions. The nuance is that now we won’t be able to easily compute the current height. It may be necessary to store the current height separately

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/a3f49d412a3c597432703bbc742245fdcba7cb64

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,178 @@
+/// Minimum number of actions required in an Orchard bundle per the Orchard protocol.
+/// The Orchard builder automatically pads bundles to meet this minimum for privacy.
+/// See: https://github.com/zcash/orchard/blob/main/src/builder.rs#L36
+pub const MIN_ACTIONS: usize = 2;
```
It looks like we still don’t have enough gas to process 2 actions. Maybe we should switch back to just one? The mempool accepts such transactions.

https://testnet.nearblocks.io/txns/9b8KqNvn7Ad5gpbx91QLa5uFVdmEiTA6fMk8r3WTM2L3?tab=execution

olga24912 replied:
fixed

olga24912 committed:
Min Actions = 1

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,178 @@
+    let ovk = orchard::keys::OutgoingViewingKey::from(BRIDGE_OVK);
+
+    for action_idx in 0..bundle.actions().len() {
+        if let Some((note, addr, _memo)) = bundle.recover_output_with_ovk(action_idx, &ovk) {
```
Recovery is a super expensive operation; we can’t afford to do it twice.

olga24912 committed:
safe output

olga24912 committed:
remove unused function

olga24912 committed:
separate validation

olga24912 committed:
merge checks

olga24912 committed:
fix add extra amount

olga24912 committed:
remove unused function

olga24912 committed:
remove changes

olga24912 committed:
fmt

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -34,7 +35,7 @@ macro_rules! define_rbf_callback {
                 #[callback_unwrap] last_block_height: u32,
             ) {
-                let expiry_height = last_block_height + self.get_config().expiry_height_gap;
+                let expiry_height = 0; //last_block_height + self.get_config().expiry_height_gap;
```
Provide expiry height in transaction init

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/21cf32667d6a60f4ec6dfce301f9aab3716e8b9f

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -96,26 +99,57 @@ impl Contract {
+        let gas_fee = if let Some(max_fee) = max_gas_fee {
+            std::cmp::min(max_fee.0, computed_gas_fee)
+        } else {
+            computed_gas_fee
```
We can remove this check

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/4a57e5b683d0b73ecfaa77764badfb5f84de81aa

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -96,26 +99,57 @@ impl Contract {
+
+            let expected_recipient = target_btc_address.clone();
+            let expected_amount = orchard_amount;
+            psbt.validate_orchard_bundle(
```
move this check inside psbt validation.

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/837d51b01ec4ce9bf043d859262b94dbb49414de

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -167,6 +204,17 @@ impl Contract {
+        let chain = self.internal_config().chain.clone();
+        if orchard_policy::has_orchard_receiver(&target_btc_address, &chain) {
+            require!(
+                orchard_bundle.is_some(),
```
This check just needs to be removed. It’s incorrect.

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/2ab5a5a8bc80417ce81b341c99fb6df3605b3b24

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,136 @@
+
+/// Check if a Unified Address contains an Orchard receiver.
+/// Returns true if the address is a Unified Address with an Orchard receiver.
+pub fn has_orchard_receiver(target_addr: &str, chain: &network::Chain) -> bool {
```
Do we need this function?

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/2ab5a5a8bc80417ce81b341c99fb6df3605b3b24

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,136 @@
+    let expected_max = expected_amount;
+    let expected_min = expected_amount.saturating_sub(min_change_amount);
+
+    require!(
```
Don't need this check

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/7e64d5af5f9f1700d294a1a4a6fb83ce9a8335dd

olga24912 commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,136 @@
+    let expected_value_balance =
+        -i64::try_from(recovered_amount).expect("Orchard amount too large for i64");
+
+    let actual_value_balance: i64 = (*value_balance).into();
```
Why i64?

olga24912 committed:
remove amount checks

olga24912 committed:
remove no needed check

olga24912 committed:
move orchard validation into check function

olga24912 committed:
remove gas calculation

olga24912 committed:
provide expire height

olga24912 committed:
fix get_branch_id

olga24912 committed:
reduce copypast

olga24912 committed:
event beckword compatibility

olga24912 committed:
fix warnings

olga24912 committed:
clippy

olga24912 committed:
fix compilation for bitcoin

olga24912 committed:
fmt

r-near committed:
fix: add expiry_height field to tests

Add missing expiry_height: None to TokenReceiverMessage::Withdraw
in test files after the field was added to the struct.

olga24912 committed:
save target_address

karim-en commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -15,6 +15,7 @@ pub enum TokenReceiverMessage {
         output: Vec<TxOut>,
         max_gas_fee: Option<U128>,
         orchard_bundle_bytes: Option<String>,
+        expiry_height: Option<u32>,
```
The `expiry_height` can be combined with `orchard_bundle_bytes` under one structure

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/9395c2297453fcc6d5b0da5a628579b9cba49484

karim-en commented on contracts/satoshi-bridge/src/api/token_receiver.rs:
```diff
@@ -15,6 +15,7 @@ pub enum TokenReceiverMessage {
         output: Vec<TxOut>,
         max_gas_fee: Option<U128>,
         orchard_bundle_bytes: Option<String>,
```
We can use [Base64VecU8](https://docs.rs/near-sdk/latest/near_sdk/json_types/struct.Base64VecU8.html) type for this.

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/7abcc1462cda80f8dc1a784ab0d36e163c24f40f

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,87 @@
+        _ => near_sdk::env::panic_str("Expected Unified Zcash address for Orchard withdraw"),
+    };
+
+    for recv in ua.items_as_parsed() {
```
Doesn't this duplicate the logic of address analysis? [network.rs](https://github.com/Near-One/btc-bridge/blob/7418581aea9ba25d7fc8a27d141ef21f4edb31ba/contracts/satoshi-bridge/src/network.rs#L171)?

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/changes/15773302fd2684e846d4e866252405155caa9a86

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -64,35 +71,74 @@ impl PsbtWrapper {
             let mut reader = Cursor::new(orchard_bundle_bytes);
-            read_v5_bundle(&mut reader).unwrap()
+            let bundle = read_v5_bundle(&mut reader)
+                .expect("Failed to read orchard bundle")
```
it is better to return result instead of panic in this function

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/4597ff0c8b263daf5b7131a3ba54ce7c6e8f5e93

In PSBT Wrapper we aren’t returning Result anywhere. I suggest fixing this in a separate PR.

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -64,35 +71,74 @@ impl PsbtWrapper {
+
+            // If no expected values provided, enforce minimum actions per Orchard protocol
+            require!(
+                bundle.actions().len() >= orchard_policy::MIN_ACTIONS,
```
Isn't better to fix it to one? otherwise more unknown actions can be added which increase the security risks?

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/d6f6b9a629f78be37f327ce59c6856630aaaea24

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -190,6 +262,27 @@ impl PsbtWrapper {
+            .unwrap();
+        } else {
+            // Write empty bundle marker
+            zcash_primitives::transaction::components::orchard::write_v5_bundle(None, &mut buf)
```
Why we need this?

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/92313d080b9a5910325e6e0dbf8022fc0489f227

karim-en commented on contracts/satoshi-bridge/src/legacy.rs:
```diff
@@ -247,3 +247,123 @@ impl From<ContractDataV1> for ContractData {
+            acc_protocol_fee_for_gas,
+        } = c;
+
+        let keys: Vec<String> = btc_pending_infos.keys().map(|k| k.clone()).collect();
```
Are you sure that we need this kind of migration? 
As I see the `recipient_address` can be included in `PsbtWrapper` which is pretty flexible. also it is needed just in the context of Zcash and psbt verification.

olga24912 replied:
fixed: 6b9d47d4fd9b9ecb9b059e488373256e319d9735

olga24912 committed:
orchard bundle in RBF

olga24912 committed:
chain specific data

olga24912 committed:
String -> Base64

olga24912 committed:
remove dublicate code

olga24912 committed:
return Error on Orchard Bundle

olga24912 committed:
check the expected actions num

olga24912 committed:
clean up

olga24912 committed:
save recipient_address to psbt

olga24912 committed:
check version

olga24912 committed:
fix tests compilation

r-near committed:
fix: update fixtures, exports, and tests

r-near committed:
fix: replace unwrap() with descriptive error messages

- bitcoin_utils/psbt_wrapper: add ERR_INVALID_PSBT_HEX, ERR_EXTRACT_TX,
  ERR_MISSING_WITNESS_UTXO, ERR_SIGHASH messages
- zcash_utils/psbt_wrapper: add ERR_INVALID_ORCHARD_BUNDLE,
  ERR_ORCHARD_VALIDATION, and detailed ERR_INVALID_PSBT messages for
  all deserialization steps
- config: add ERR_CONFIG messages for uninitialized contract/config
- account: include account_id in ERR_ACCOUNT_NOT_REGISTERED message

olga24912 commented on contracts/satoshi-bridge/src/api/bridge.rs:
```diff
@@ -116,7 +116,13 @@ impl Contract {
     /// * `output` - Modified output.
     #[pause(except(roles(Role::DAO)))]
-    pub fn withdraw_rbf(&mut self, original_btc_pending_verify_id: String, output: Vec<TxOut>) {
+    pub fn withdraw_rbf(
```
We should add extra public function, where we can just fix bundle with same parameters.

karim-en commented on contracts/satoshi-bridge/src/bitcoin_utils/contract_methods.rs:
```diff
@@ -11,7 +12,14 @@ macro_rules! define_rbf_method {
             original_btc_pending_verify_id: String,
             output: Vec<TxOut>,
+            orchard_bundle_bytes: Option<String>,
+            expiry_height: Option<u32>,
```
Let's use `ChainSpecificData`

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/ed71a2f7043a8584b15a6cb9044a3b181989417f

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -15,11 +16,19 @@ macro_rules! define_rbf_callback {
                 original_btc_pending_verify_id: String,
                 output: Vec<TxOut>,
+                orchard_bundle_bytes: Option<String>,
+                expiry_height: Option<u32>,
```
Let's use `ChainSpecificData`

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/ed71a2f7043a8584b15a6cb9044a3b181989417f

karim-en commented on contracts/satoshi-bridge/src/zcash_utils/types.rs:
```diff
@@ -0,0 +1,8 @@
+#[near(serializers = [json])]
+pub struct ChainSpecificData {
+    pub orchard_bundle_bytes: Option<Base64VecU8>,
+    pub expiry_height: Option<u32>,
```
We can remove the `Option` here

olga24912 replied:
fixed: https://github.com/Near-One/btc-bridge/pull/20/commits/ed71a2f7043a8584b15a6cb9044a3b181989417f

olga24912 committed:
provide chain_specific data

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/contract_methods.rs:
```diff
@@ -93,17 +119,40 @@ impl Contract {
-            orchard_bundle,
+            input.clone(),
+            output.clone(),
+            orchard_bundle.clone(),
```
`input`, `output` and `orchard_bundle` are not used after this, so `.clone()` can be dropped and we can move them

r-near replied:
Fixed [`1fa5468` (#20)](https://github.com/Near-One/btc-bridge/pull/20/commits/1fa5468134aa5b840a6291aa9902fb110a69dd0c)

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,106 @@
+) -> Result<
+    (
+        Option<orchard::Bundle<orchard::bundle::Authorized, ZatBalance>>,
+        Option<(u64, [u8; 43])>,
```
This types looks too `magical` maybe it's better to create an alias or at least pin `43` with some constant

r-near replied:
Added `ORCHARD_RAW_ADDRESS_SIZE` constant and `OrchardRawAddress` type alias https://github.com/Near-One/btc-bridge/pull/20/changes/3ab32bb041005ab3bd1297a153029fca499e3fc9

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        // If no expected values provided, enforce minimum actions per Orchard protocol
+        if bundle.actions().len() != EXPECTED_ACTIONS_NUMBER {
```
This check can be performed at the very beginning

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/commit/3ab32bb041005ab3bd1297a153029fca499e3fc9

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,106 @@
+
+        let ovk = orchard::keys::OutgoingViewingKey::from(BRIDGE_OVK);
+
+        for action_idx in 0..bundle.actions().len() {
```
loop can be dropped entirely, since `bundle.actions().len()` must be 1 (or it can be kept if you're planning to bump `EXPECTED_ACTIONS_NUMBER` to 2 in the future)

r-near replied:
Good point. We can't fit 2 actions in without going over on our gas, so better to simplify. Fixed it here https://github.com/Near-One/btc-bridge/commit/3ab32bb041005ab3bd1297a153029fca499e3fc9

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/orchard_policy.rs:
```diff
@@ -0,0 +1,106 @@
+    bundle: &Bundle<orchard::bundle::Authorized, ZatBalance>,
+    expected_recipient: &str,
+    chain: &network::Chain,
+    orchard_output: (u64, [u8; 43]),
```
same here with alias

r-near replied:
Applied the type alias here https://github.com/Near-One/btc-bridge/commit/3ab32bb041005ab3bd1297a153029fca499e3fc9

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -24,6 +24,8 @@ pub struct PsbtWrapper {
     vout: Vec<ZcashTxOut>,
     inputs_utxo: Vec<ZcashTxOut>,
     orchard_bundle: Option<orchard::Bundle<orchard::bundle::Authorized, ZatBalance>>,
+    orchard_output: Option<(u64, [u8; 43])>,
```
same here with alias

r-near replied:
Fixed https://github.com/Near-One/btc-bridge/commit/3ab32bb041005ab3bd1297a153029fca499e3fc9

frolvanya commented on contracts/satoshi-bridge/src/zcash_utils/psbt_wrapper.rs:
```diff
@@ -190,49 +233,103 @@ impl PsbtWrapper {
 
     pub fn deserialize(psbt_hex: &String) -> Self {
-        let bytes = hex::decode(psbt_hex).unwrap();
+        let bytes = hex::decode(psbt_hex).expect("ERR_INVALID_PSBT_HEX: failed to decode hex");
```
In the future it might be worth to change all `expect` to `env::panic_str` to reduce contract size

r-near replied:
Fixed [`051db31` (#20)](https://github.com/Near-One/btc-bridge/pull/20/commits/051db31891e0f4c32ef41ae34f2bc6e7c35d8cd2)

frolvanya replied:
I think that that there should be a lot more places with `expect` in other files, since satoshi used it for some reason. But I think that can be a job for another PR

r-near committed:
remove unnecessary .clone() calls

r-near committed:
add type alias for Orchard raw address bytes and simplify extract_orchard_bundle

r-near committed:
replace expect() with env::panic_str() to reduce contract size

frolvanya commented on contracts/satoshi-bridge/src/network.rs:
```diff
@@ -200,6 +200,24 @@ impl Address {
     }
 
+    /// Extract the Orchard receiver raw bytes from a Unified Address string for the given chain.
+    pub fn extract_orchard_receiver(&self) -> Result<[u8; 43], String> {
```
```suggestion
    pub fn extract_orchard_receiver(&self) -> Result<OrchardRawAddress, String> {
```

frolvanya replied:
Fixed in https://github.com/Near-One/btc-bridge/pull/20/commits/bc5582ff005b2d933a032a1f2732a04738da6f5f + 0e43effabc296fbbcf9976222b80a84292ba52eb

r-near committed:
Update contracts/satoshi-bridge/src/network.rs

Co-authored-by: Ivan Frolov <59515280+frolvanya@users.noreply.github.com>

frolvanya committed:
chore: imported `OrchardRawAddress`
